configuration/mongo-configuration.ts

```diff
@@ -0,0 +1,124 @@
+import {MongoClient, Db, Collection} from "mongodb";
+import {IConfiguration} from "./iconfiguration";
+
+export class MongoConfiguration implements IConfiguration {
+    /**
+     * Number of seconds to wait between each attempt at connecting
+     * to the DB.
+     */
+    private static timeoutSeconds: number = 8;
+    private mongoConfig: { [key: string]: any } = {};
+    private requiredKeys: string[];
+    private secondsToRetry: number;
+    private mongoUri: string;
+    private db: Db;
+    private collection: Collection;
+
+    /**
+     * Asynchronously initialize configuration values from the passed
+     * Mongo database.
+     *
+     * Retries the DB connection up to secondsToRetry seconds. Returns early if
+     * connection is successful and requiredKeys are found in the
+     * collectionName collection.
+     *
+     * @param {string} mongoUri - URI of the desired Mongo DB.
+     * @param {string} collectionName - Specific collection used to store/get
+     * values.
+     * @param {string[]} requiredKeys - Array of keys required to be contained
+     * in the collectionName collection before completing initialization.
+     * @param {number} secondsToRetry - Max number of seconds to wait for
+     * a connection and all requiredKeys before throwing an error.
+     */
+    public async initialize(
+            mongoUri: string,
+            collectionName: string = "config",
+            requiredKeys: string[] = [],
+            secondsToRetry: number = 5 * 60): Promise<void> {
+        this.requiredKeys = requiredKeys;
+        this.secondsToRetry = secondsToRetry;
+        this.mongoUri = mongoUri;
+        await this.initializeDb();
+
+        this.collection = this.db.collection(collectionName);
+        this.mongoConfig = await this.collection.findOne({});
+        return undefined;
+    }
+
+    private async initializeDb(): Promise<void> {
+        let numRetries = this.secondsToRetry / MongoConfiguration.timeoutSeconds;
```

Comment:
Instead of keeping track of numRetries, keep track of the time since the function was initially started and compare against secondsToRetry directly.